src/Common/MemoryWorker.cpp

```diff
@@ -0,0 +1,49 @@
+#include <Common/MemoryWorker.h>
+
+#include <Common/Jemalloc.h>
+#include <Common/MemoryTracker.h>
+#include <Common/formatReadable.h>
+
+namespace DB
+{
+
+#if USE_JEMALLOC
+MemoryWorker::MemoryWorker(uint64_t period_ms_)
+    : period_ms(period_ms_)
+{
+    background_thread = ThreadFromGlobalPool([this] { backgroundThread(); });
+}
+
+MemoryWorker::~MemoryWorker()
+{
+    {
+        std::unique_lock lock(mutex);
+        shutdown = true;
+    }
+    cv.notify_all();
+
+    if (background_thread.joinable())
+        background_thread.join();
+}
+
+void MemoryWorker::backgroundThread()
+{
+    JemallocMibCache<uint64_t> epoch_mib("epoch");
+    JemallocMibCache<size_t> resident_mib("stats.resident");
+    std::unique_lock lock(mutex);
+    while (true)
+    {
+        cv.wait_for(lock, period_ms, [this] { return shutdown; });
+        if (shutdown)
+            return;
+
+        epoch_mib.setValue(0);
+        Int64 resident = resident_mib.getValue();
+        MemoryTracker::setRSS(resident);
```

Comment:
Optimistically, we can try to avoid doing this at all. The only scary thing is that we can account memory incorrectly sometimes, and even small bug can give a big drift with time. But maybe we can try anyway (under the option).

If we afraid OOMs, I would better check two counters at the same time for total tracker (our internal counted memory usage; last visible RSS + allocations delta).

Reply:
The problem isn't the drift because of wrong accounting but jemalloc's dirty pages which can be couple of GiBs.
Disabling them can affect performance so we have to account for them somehow.
`1s` in AsyncMetrics wasn't enough to avoid OOM so something more frequent needs to trigger.
This thread should be really lightweight as we control the update with `epoch` and fetch a single stat (everything is in-memory).
Is there a reason why we don't want to update the total tracker with RSS? What do we get with two trackers?

Reply:
The issue could be the following:
* Run a query which allocates and then deallocates a lot of memory
* Because of some reasons (jemalloc cache, or some natural race when rss accounted deallocations later than our MemoryTracker counter), visible RSS is still high and won't be updated till the next period
* The next query can't allocate the memory even if it could.

I think it's just wrong to set a global memory tracker counter from RSS, because of the race condition, and because they count a little bit different things.

There could be another issue (which @amosbird told me about) about page faulting: in case we allocate a big chunk of memory, it may not be fully visible in RSS until all the pages are faulted. And it looks like there is nothing we can do with it (memory is accounted for in MemoryTracker but still not visible in RSS, which value we can take). I do not know whether it actually happens or not, but it at least looks plausible.

Two separate counters would allow at least not mixing two different metrics.

Reply:
As long as we don't allocate directly using mmap, I would trust metrics from jemalloc more than metrics from MemoryTracker.
(fast search gave me those, hope they are not important
https://github.com/unicode-org/icu/blob/a56dde820dc35665a66f2e9ee8ba58e75049b668/icu4c/source/common/umapfile.cpp#L239
https://github.com/unum-cloud/usearch/blob/955c6f9c11adfd89c912e0d1643d160b4e9e543f/cpp/bench.cpp#L115
https://github.com/ClickHouse/ClickHouse/blob/80e1377e5d2223176274c319938187f0da799280/src/Common/PageCache.cpp#L621
https://github.com/ClickHouse/libhdfs3/blob/0d04201c45359f0d0701fb1e8297d25eff7cfecf/src/common/MappedFileWrapper.cpp#L59
https://github.com/intel/qpl/blob/d4715e0e79896b85612158e135ee1a85f3b3e04d/sources/middle-layer/dispatcher/hw_queue.cpp#L113
)
Maybe we can try checking `stats.allocated` vs MemoryTracker and try to detect big differences...

Reply:
> As long as we don't allocate directly using mmap

that was my question too. there are some other ways i believe how memory might be allocated bypassing jemalloc. what libraries do we don't even have a clue. so it seems a little fragile to rely on the allocator. especially when we have access to the system's view on our memory usage (at the end this metric will be used by oom killer and other memory reclamation routines).

Reply:
That is understandable, and we can easily replace in MemoryWorker source of resident memory to something else (we can keep `stats.allocated` to correct MemoryTracker in extreme cases)
But there is still a tradeoff of having RSS as up-to-date as possible vs covering everything.

The more important thing is updating more often than every 1s, reading from jemalloc just seems the most lightweight.